### PR TITLE
Remove AccountUpdateProfilePage from the testsuite

### LINKED ISSUE #15202: Remove AccountUpdateProfilePage from the testsuite
### Description

Tests that are not testing the account console should not rely on it. This task is to remove [AccountUpdateProfilePage](https://github.com/keycloak/keycloak/blob/3af1134975f8cb93264035fb64edce953e2e7bbf/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/AccountUpdateProfilePage.java) and usage of it, re-writing to use REST APIs as applicable.

### LINKED ISSUE #15203: Update tests to not rely on old account console for login/logout
### Description

There's a few tests that rely on the old account console to test users log in or log out. These should be updated to use the testing application, or some other means to log in or log out. 

The tests should not be updated to rely on the new account console!

### PR DESCRIPTION
**Pull request by Aboullos**
Closes #15202 and partially closes keycloak#15203 and keycloak#19668

### CODE REVIEW

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/AccountHelper.java [hunk 1]
```diff
@@ -44,4 +59,16 @@ public static boolean updatePassword(RealmResource realm, String username, Strin
         }
     }
 
+    public static String getEmail(RealmResource realm, String username) {
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/AccountHelper.java:62]: I think these methods are too fine-grained for a generic `AccountHelper` class. Each of them is calling `getUserRepresentation`, which performs a search query. If you need email, first name and last name in a test, you are querying the DB 3 times. 

I think it's better to use `getUserRepresentation` directly in the test method.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/AccountHelper.java [hunk 2]
```diff
@@ -25,14 +25,29 @@
...
-        UserResource user = realm.users().get(userRepresentation.getId());
+        return userResult.get();
+    }
+
+    public static void logout(RealmResource realm, String username) {
+        realm.users().get(getUserRepresentation(realm, username).getId()).logout();
+    }
+
+    public static String getUsername(RealmResource realm, String username) {
+        return getUserRepresentation(realm, username).getUsername();
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/AccountHelper.java:42]: What is the difference between the parameter `username` and the returned in `getUsername`? Isn't it the same?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java [hunk 3]
```diff
@@ -351,7 +348,6 @@ public void loginDisabledUser() {
     }
 
     @Test
-    @DisableFeature(value = Profile.Feature.ACCOUNT2, skipRestart = true) // TODO remove this (KEYCLOAK-16228)
     public void loginDifferentUserAfterDisabledUserThrownOut() {
         String userId = adminClient.realm("test").users().search("test-user@localhost").get(0).getId();
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java:352]: Here you can use `AccountHelper.getUserRepresentation().getId()`

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/AccountHelper.java [hunk 4]
```diff
@@ -25,14 +25,29 @@
...
+    }
+
+    public static void updateUsername(RealmResource realm, String username, String newUsername) {
+        getUserRepresentation(realm, username).setUsername(newUsername);
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/AccountHelper.java:46]: Is this setter actually updating the object in the DB?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java [hunk 5]
```diff
@@ -363,19 +359,19 @@ public void loginDifferentUserAfterDisabledUserThrownOut() {
             appPage.assertCurrent();
             appPage.openAccount();
 
-            profilePage.assertCurrent();
-
             setUserEnabled(userId, false);
 
             // force refresh token which results in redirecting to login page
-            profilePage.updateUsername("notPermitted");
+            AccountHelper.updateUsername(adminClient.realm("test"), "test-user@localhost", "notPermitted");
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java:365]: Not sure this is actually doing anything relevant. Previously it was updating the username only for forcing the refresh token. This caused a redirect to login page. Now you are just setting a new username and manually opening the login page, with no refresh token update.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/custom/CustomThemeTest.java [hunk 6]
```diff
@@ -61,26 +58,25 @@ public void configureTestRealm(RealmRepresentation testRealm) {
     protected LoginPage loginPage;
 
     @Page
-    protected AccountUpdateProfilePage profilePage;
+    protected AppPage appPage;
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/custom/CustomThemeTest.java:61]: Is `appPage` used anywhere in this or any child class?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/custom/CustomThemeTest.java [hunk 7]
```diff
@@ -61,26 +58,25 @@ public void configureTestRealm(RealmRepresentation testRealm) {
...
+        UserRepresentation userRep = AccountHelper.getUserRepresentation(testRealm(), "test-user@localhost");
 
-        events.expectLogin().client("account").detail(Details.REDIRECT_URI, getAccountRedirectUrl()).assertEvent();
+        Assert.assertEquals("test-user@localhost", userRep.getEmail());
+        Assert.assertNull(userRep.firstAttribute("street"));
 
-        Assert.assertEquals("test-user@localhost", profilePage.getEmail());
-        Assert.assertEquals("", profilePage.getAttribute("street"));
+        userRep.singleAttribute("street", "Elm 1");
+        testRealm().users().get(userRep.getId()).update(userRep);
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/custom/CustomThemeTest.java:72]: After updating the `userRep` you have to re-fetch it from the DB for asserting the correct values.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractBaseBrokerTest.java [hunk 8]
```diff
@@ -406,10 +403,11 @@ protected void assertLoggedInAccountManagement() {
     }
 
     protected void assertLoggedInAccountManagement(String username, String email) {
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractBaseBrokerTest.java:405]: Does it make sense to keep this method? Can we remove it? What is it good for?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractFirstBrokerLoginTest.java [hunk 9]
```diff
@@ -581,11 +582,11 @@ public void testFixDuplicationsByReviewProfile() {
...
-        accountUpdateProfilePage.assertCurrent();
-        Assert.assertEquals("FirstName", accountUpdateProfilePage.getFirstName());
-        Assert.assertEquals("LastName", accountUpdateProfilePage.getLastName());
-        Assert.assertEquals("test@localhost.com", accountUpdateProfilePage.getEmail());
-        Assert.assertEquals("test", accountUpdateProfilePage.getUsername());
+
+        Assert.assertEquals("FirstName", AccountHelper.getFirstName(adminClient.realm(bc.consumerRealmName()), "test"));
+        Assert.assertEquals("LastName", AccountHelper.getLastName(adminClient.realm(bc.consumerRealmName()), "test"));
+        Assert.assertEquals("test@localhost.com", AccountHelper.getEmail(adminClient.realm(bc.consumerRealmName()), "test"));
+        Assert.assertEquals("test", AccountHelper.getUsername(adminClient.realm(bc.consumerRealmName()), "test"));
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractFirstBrokerLoginTest.java:589]: Get user representation once.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractFirstBrokerLoginTest.java [hunk 10]
```diff
@@ -1001,11 +996,11 @@ public void testEventsOnUpdateProfileNoEmailChange() {
...
-        accountUpdateProfilePage.assertCurrent();
-        Assert.assertEquals("FirstName", accountUpdateProfilePage.getFirstName());
-        Assert.assertEquals("LastName", accountUpdateProfilePage.getLastName());
-        Assert.assertEquals("no-first-name@localhost.com", accountUpdateProfilePage.getEmail());
-        Assert.assertEquals("no-first-name", accountUpdateProfilePage.getUsername());
+
+        Assert.assertEquals("FirstName", AccountHelper.getFirstName(adminClient.realm(bc.consumerRealmName()), "no-first-name"));
+        Assert.assertEquals("LastName", AccountHelper.getLastName(adminClient.realm(bc.consumerRealmName()), "no-first-name"));
+        Assert.assertEquals("no-first-name@localhost.com", AccountHelper.getEmail(adminClient.realm(bc.consumerRealmName()), "no-first-name"));
+        Assert.assertEquals("no-first-name", AccountHelper.getUsername(adminClient.realm(bc.consumerRealmName()), "no-first-name"));
```
Comment (miquelsi) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractFirstBrokerLoginTest.java:1003]: Get userRep once.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/AccountHelper.java (hunk 1 above)
Comment (Aboullos) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/AccountHelper.java:62]: There are 14 usages of getEmail(), and 15 of getFirstName() and getLastName(). It is for reusability of the code and easy the modification of it

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java (hunk 5 above)
Comment (Aboullos) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java:365]: Yes, but I thought the updateUsername on the API also refreshed the token. If it is not like that I'll remove it

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/custom/CustomThemeTest.java (hunk 6 above)
Comment (Aboullos) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/custom/CustomThemeTest.java:61]: Neither that one nor `loginPage`, removing them

**Review by miquelsi — DISMISSED** [review]

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/custom/CustomThemeTest.java
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/custom/CustomThemeTest.java:1]: This should just be removed as it's not serving any purpose now.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/ClientInitiatedAccountLinkTest.java [hunk 11]
```diff
@@ -560,12 +555,11 @@ public void testAccountLinkingExpired() throws Exception {
         Assert.assertTrue(links.isEmpty());
 
         // Login to account mgmt first
-        profilePage.open(CHILD_IDP);
+        driver.navigate().to(RealmsResource.accountUrl(UriBuilder.fromUri(getAuthServerRoot())).build(CHILD_IDP).toString());
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/ClientInitiatedAccountLinkTest.java:558]: We shouldn't navigate to the account url here, as that will open the new admin console. I'd change this to use the test app instead.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractAdvancedBrokerTest.java [hunk 12]
```diff
@@ -157,7 +156,7 @@ public void testAccountManagementLinkedIdentityAlreadyExists() {
         driver.navigate().to(getAccountUrl(getConsumerRoot(), bc.consumerRealmName()));
         logInWithBroker(bc);
         waitForAccountManagementTitle();
-        accountUpdateProfilePage.assertCurrent();
+        Assert.assertFalse(loginPage.isCurrent());
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractAdvancedBrokerTest.java:159]: Should we updated to login to test app, and check that is opened rather the checking login page isn't current page

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractAdvancedBrokerTest.java [hunk 13]
```diff
@@ -190,7 +189,7 @@ public void testRetrieveToken() throws Exception {
         logInWithBroker(bc);
         updatePasswordPage.updatePasswords("password", "password");
         waitForAccountManagementTitle();
-        accountUpdateProfilePage.assertCurrent();
+        Assert.assertFalse(loginPage.isCurrent());
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractAdvancedBrokerTest.java:192]: Should we updated to login to test app, and check that is opened rather the checking login page isn't current page

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractAdvancedBrokerTest.java [hunk 14]
```diff
@@ -157,7 +156,7 @@ public void testAccountManagementLinkedIdentityAlreadyExists() {
         driver.navigate().to(getAccountUrl(getConsumerRoot(), bc.consumerRealmName()));
         logInWithBroker(bc);
         waitForAccountManagementTitle();
-        accountUpdateProfilePage.assertCurrent();
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractAdvancedBrokerTest.java:160]: This will need to be updated to not rely on the account console at all, but I presume that will be handled in a follow-up?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractBaseBrokerTest.java [hunk 15]
```diff
@@ -263,7 +256,7 @@ protected void logInAsUserInIDPForFirstTime() {
 
     protected void logInAsUserInIDPForFirstTimeAndAssertSuccess() {
         logInAsUserInIDPForFirstTime();
-        assertLoggedInAccountManagement();
+        Assert.assertFalse(loginPage.isCurrent());
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractBaseBrokerTest.java:259]: Shouldn't this be updated to login to test app, and check that is opened rather the checking login page isn't current page

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractFirstBrokerLoginTest.java [hunk 16]
```diff
@@ -133,7 +134,7 @@ public void testLinkAccountByReauthenticationWithPassword() {
 
         loginPage.login("password");
         waitForAccountManagementTitle();
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractFirstBrokerLoginTest.java:136]: Shouldn't this be updated to login to test app, and check that is opened rather the checking login page isn't current page

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractGroupMapperTest.java [hunk 17]
```diff
@@ -86,7 +87,7 @@ protected UserRepresentation loginAsUserTwiceWithMapper(
         updateUser();
 
         logInAsUserInIDP();
-        assertLoggedInAccountManagement();
+        Assert.assertFalse(loginPage.isCurrent());
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractGroupMapperTest.java:90]: Shouldn't this be updated to login to test app, and check that is opened rather the checking login page isn't current page

**Comment by stianst** [comment]
Was going through this and there's still some places that is leveraging the account console, which I presume we can handle in a follow-up, but I can see some places here that is now not checking if login passed, but rather just the login page isn't the current page.

**Review by mposolda — CHANGES_REQUESTED** [review]
I've added few inline comments.

Besides that, I agree with @stianst that `assertFalse(loginPage.isCurrent());` is quite bad way to check if user is really authenticated. It will be nice if those tests are changed to use `AppPage` and the example application when possible and check that logged-in user in similar way like for example `LoginTest` is doing. Or eventually some other page similar to AppPage when it makes sense. I hope it is possible to change most of the places in this PR this way?

**Review by miquelsi — DISMISSED** [review]

**Review by mposolda — CHANGES_REQUESTED** [review]
@Aboullos I see that most of the conversations, which were pointed by @stianst about not using:
```
Assert.assertFalse(loginPage.isCurrent());
```
were resolved. However I also see that nothing is changed and those are still used as before :-)

Isn't it really possible to rather directly check the page where the browser is? That's better than checking "not being on the login page", which is quite fragile (as the fact you are not on the login page doesn't mean that you are still in the middle of something in the login process, for example some requiredAction page etc)? 


In other words, use something like `appPage.assertCurrent()` or `someBrokerTestAppPage.assertCurrent()` instead of  `Assert.assertFalse(loginPage.isCurrent());`.

WDYT?

I am marking this as "Request changes" until we clarify this.

**Comment by Aboullos** [comment]
> @Aboullos I see that most of the conversations, which were pointed by @stianst about not using:
> 
> ```
> Assert.assertFalse(loginPage.isCurrent());
> ```
> 
> were resolved. However I also see that nothing is changed and those are still used as before :-)
> 
> Isn't it really possible to rather directly check the page where the browser is? That's better than checking "not being on the login page", which is quite fragile (as the fact you are not on the login page doesn't mean that you are still in the middle of something in the login process, for example some requiredAction page etc)?
> 
> In other words, use something like `appPage.assertCurrent()` or `someBrokerTestAppPage.assertCurrent()` instead of `Assert.assertFalse(loginPage.isCurrent());`.
> 
> WDYT?
> 
> I am marking this as "Request changes" until we clarify this.

Sorry, I did a revert in some classes and some of the asserts went back, but there were still some changed to appPage.assertCurrent. Just checked all the changes and updated the assertFalse of the login page to appPage.assertCurrent

**Comment by mposolda** [comment]
@Aboullos Thanks for the update! The changes looks good to me in this PR.

There is failing test. Could you check it? Maybe just rebase this PR on top of latest main will help with failing test? Anyway, when the failing test is addressed, you can consider this PR as approved from me :-)

**Review by mposolda — APPROVED** [review]

**Comment by mposolda** [comment]
@Aboullos Approving now. Thanks for all the updates and hard work on this PR! 
@miquelsi Let me know if you want to merge this PR. I can do that whenever I have green light from you (or feel free to merge it yourself).

**Review by miquelsi — APPROVED** [review]
